File: src/mbs/application/common/article_gateway.py
from abc import abstractmethod
from typing import Protocol

from mbs.domain.models import Article, ArticleId, ArticleState, UserId, File, FileId


class ArticleReader(Protocol):
    @abstractmethod
    async def get_article(self, article_id: ArticleId) -> Article | None:
        pass

    @abstractmethod
    async def get_articles(
            self,
            limit: int,
            offset: int,
            order_by: str,
            state: ArticleState,
            tag: str = None,
            query: str = None,
            author_id: UserId = None,
    ) -> list[Article]:
        pass


class ArticleWriter(Protocol):
    @abstractmethod
    async def save_article(self, article: Article) -> None:
        pass


class ArticleRater(Protocol):
    @abstractmethod
    async def rate_article(self, article_id: ArticleId, user_id: UserId) -> None:
        """
        Rate article by user

        if user already rated article, then remove rating
        """
        pass

    @abstractmethod
    async def is_article_rated(self, article_id: ArticleId, user_id: UserId) -> bool:
        pass

    @abstractmethod
    async def is_articles_rated(self, article_ids: list[ArticleId], user_id: UserId) -> list[bool]:
        pass


class ArticleFile(Protocol):
    @abstractmethod
    async def get_article_files(self, article_id: ArticleId) -> list[File]:
        pass

    @abstractmethod
    async def is_file_linked_to_article(self, article_id: ArticleId, file_id: FileId) -> bool:
        pass

    @abstractmethod
    async def link_file_to_article(self, article_id: ArticleId, file_id: FileId) -> None:
        pass

    @abstractmethod
    async def unlink_file_from_article(self, article_id: ArticleId, file_id: FileId) -> None:
        pass


class ArticleRemover(Protocol):
    @abstractmethod
    async def remove_article(self, article_id: ArticleId) -> None:
        pass

File: src/application/common/article_gateway.rs
use async_trait::async_trait;
use crate::application::article::find::OrderBy;
use crate::domain::models::{
    article::{Article, ArticleId},
    article_state::ArticleState,
    tag::TagId,
    user_id::UserId
};
use crate::domain::models::rate_state::RateState;

pub enum ArticleGatewayError {
    Critical(String)
}

#[async_trait]
pub trait ArticleReader {
    async fn get_article(&self, id: &ArticleId) -> Result<Option<Article>, ArticleGatewayError>;
    async fn find_articles(
        &self,
        query: Option<String>,
        limit: u8,
        offset: u32,
        order_by: &OrderBy,
        state: &ArticleState,
        tags: &[TagId],
        author_id: &Option<UserId>,
    ) -> Result<Vec<Article>, ArticleGatewayError>;
    async fn get_article_author(&self, article_id: &ArticleId) -> Result<Option<UserId>, ArticleGatewayError>;
    async fn get_article_state(
        &self,
        article_id: &ArticleId,
    ) -> Result<Option<ArticleState>, ArticleGatewayError>;
}

#[async_trait]
pub trait ArticleWriter {
    async fn save(&self, article: &Article) -> Result<(), ArticleGatewayError>;
    async fn increment_article_views(&self, id: &ArticleId) -> Result<(), ArticleGatewayError>;
}

#[async_trait]
pub trait ArticleRemover {
    async fn remove(&self, article_id: &ArticleId) -> Result<(), ArticleGatewayError>;
}

#[async_trait]
pub trait ArticleRater {
    async fn rate_article(&self, article_id: &ArticleId, user_id: &UserId, state: &RateState) -> Result<(), ArticleGatewayError>;
    async fn user_rate_state(&self, article_id: &ArticleId, user_id: &UserId) -> Result<RateState, ArticleGatewayError>;
    async fn user_rate_states(
        &self,
        article_ids: &[ArticleId],
        user_id: &UserId,
    ) -> Result<Vec<RateState>, ArticleGatewayError>;
}

#[async_trait]
pub trait ArticleGateway: ArticleReader + ArticleWriter + ArticleRemover + ArticleRater {}
